cmd/fleet/main.go

```diff
@@ -84,24 +116,184 @@ func NewCommand(version string) *cobra.Command {
 		RunE:  getRunCommand(version),
 	}
 	cmd.Flags().StringP("config", "c", "fleet-server.yml", "Configuration for Fleet Server")
+	cmd.Flags().Bool("agent-mode", false, "Running under execution of the Elastic Agent")
 	return cmd
 }
 
-type FleetServer struct {
+type AgentMode struct {
+	cache   cache.Cache
 	version string
 
-	cfg   *config.Config
-	cfgCh chan *config.Config
-	cache cache.Cache
+	reloadables []reload.Reloadable
+
+	client client.Client
+
+	lock         sync.Mutex
+	firstCfg     chan *config.Config
+	srv          *FleetServer
+	srvCtx       context.Context
+	srvCanceller context.CancelFunc
+	startChan    chan struct{}
+}
+
+func NewAgentMode(reader io.Reader, c cache.Cache, version string, reloadables ...reload.Reloadable) (*AgentMode, error) {
+	var err error
+
+	a := &AgentMode{
+		cache:       c,
+		version:     version,
+		reloadables: reloadables,
+	}
+	a.client, err = client.NewFromReader(reader, a)
+	if err != nil {
+		return nil, err
+	}
+	return a, nil
+}
+
+func (a *AgentMode) Run(ctx context.Context) error {
+	ctx, canceller := context.WithCancel(ctx)
+	defer canceller()
+
+	a.firstCfg = make(chan *config.Config)
+	a.startChan = make(chan struct{}, 1)
+	log.Info().Msg("starting communication connection back to Elastic Agent")
+	err := a.client.Start(ctx)
+	if err != nil {
+		return err
+	}
+
+	// wait for the initial configuration to be sent from the
+	// Elastic Agent before starting the actual Fleet Server.
+	log.Info().Msg("waiting for Elastic Agent to send initial configuration")
+	var cfg *config.Config
+	select {
+	case <-ctx.Done():
+		return fmt.Errorf("never received initial configuration")
+	case cfg = <-a.firstCfg:
+	}
+
+	// start fleet server with the initial configuration and its
+	// own context (needed so when OnStop occurs the fleet server
+	// is stopped and not the elastic-agent-client as well)
+	srvCtx, srvCancel := context.WithCancel(ctx)
+	defer srvCancel()
+	log.Info().Msg("received initial configuration starting Fleet Server")
+	srv, err := NewFleetServer(cfg, a.cache, a.version, status.NewChained(status.NewLog(), a.client))
+	if err != nil {
+		// unblock startChan even though there was an error
+		a.startChan <- struct{}{}
+		return err
+	}
+	a.lock.Lock()
+	close(a.firstCfg)
+	a.firstCfg = nil
+	a.srv = srv
+	a.srvCtx = srvCtx
+	a.srvCanceller = srvCancel
+	a.lock.Unlock()
+
+	// trigger startChan so OnConfig can continue
+	a.startChan <- struct{}{}
+
+	return a.srv.Run(srvCtx)
+}
+
+func (a *AgentMode) OnConfig(s string) {
```

Comment:
Are we getting only fleet configuration here?  Hate to cause a server bounce on a config change elsewhere.

Reply:
Agent only sends what Fleet Server is allowed to have. We also already have code in place to bounce on changes.

Reply:
That's cool.  So agent only sends section if it changed?  Or do we get the fleet section if any of the other sections change and will that force a bounce?

Reply:
Agent resends the whole config if something changes (that is related to Fleet Server). Our internal code inside of Fleet Server ensures to only bounce things that need it based on the change.